include/seqan3/alignment/pairwise/execution/execution_handler_parallel.hpp

```diff
@@ -0,0 +1,169 @@
+            state->thread_pool.emplace_back([q] ()
+            {
+                for (;;)
+                {
+                    task_type task;
+                    if (q->wait_pop(task) == contrib::queue_op_status::closed)
+                        return;
+
+                    task();
+                }
+            });
+        }
+    }
+
+    //!\brief Constructs the execution handler with std::thread::hardware_concurrency.
+    execution_handler_parallel() : execution_handler_parallel{std::thread::hardware_concurrency()}
+    {}
+
+    execution_handler_parallel(execution_handler_parallel const &) = delete;                 //!< Deleted.
+    execution_handler_parallel(execution_handler_parallel &&) = default;                     //!< Defaulted.
+    execution_handler_parallel & operator=(execution_handler_parallel const &) = delete;     //!< Deleted.
+    execution_handler_parallel & operator=(execution_handler_parallel &&) = default;         //!< Defaulted.
+
+    //!\brief Waits for threads to finish.
+    ~execution_handler_parallel()
+    {
+        if (state != nullptr)
+            wait();
+    }
+    //!\}
+
+    /*!\name Execution
+     * \{
+     */
+    /*!\brief Invokes the passed alignment instance in a non-blocking manner.
+     * \copydetails execution_handler_sequential::execute
+     */
+
+     // * \tparam fn_type           The callable that needs to be invoked; must model std::Invocable with first_range_type
+     // *                           and second_range_type.
+     // * \tparam first_range_type  The type of the first range.
+     // * \tparam second_range_type The type of the second range.
+     // * \tparam delegate_type     The type of the callable invoked on the std::invoke_result of `fn_type`; must model
+     // *                           std::Invocable.
+     // *
+     // * \param[in] func         The callable invoking the alignment algorithm.
+     // * \param[in] idx          The index of the current processed sequence pair.
+     // * \param[in] first_range  The first range.
+     // * \param[in] second_range The second range.
+     // * \param[in] delegate     The callable invoked with the result of the alignment.
+    template <typename fn_type, typename first_range_type, typename second_range_type, typename delegate_type>
+    //!\cond
+        requires std::Invocable<fn_type, size_t const, first_range_type, second_range_type> &&
+                 std::Invocable<delegate_type, std::invoke_result_t<fn_type,
+                                                                    size_t const,
+                                                                    first_range_type,
+                                                                    second_range_type>>
+    //!\endcond
+    void execute(fn_type && func,
+                 size_t const idx,
+                 first_range_type && first_range,
+                 second_range_type && second_range,
+                 delegate_type && delegate)
```

Comment:
are the ranges here views now too?

Reply:
They are potentially always views. I could copy them instead getting them by forwarding reference.

Reply:
Then again, views might be move-only in the future...

Reply:
yes, but I always take as non-reference so the caller needs to move into this, which is by the current design ensured. I do assert that the types are indeed views.